Store balance transaction ID in order metadata

anu-rock opened the pull request:
Fixes #7856

#### Changes proposed in this Pull Request

* Add a new key `_wcpay_payment_transaction_id` in order metadata to store Stripe's balance transaction ID for each payment to aid in transaction reconciliation in third-party accounting systems.
* Refactor the `attach_intent_info_to_order` method to make it easier to add new intent-based data values to order meta.

#### Testing instructions

* Place a new order in WooCommerce frontend with any payment method.
* Make a GET request to `/wp-json/wc/v3/orders/<order_id>` (see [Retrieve an order API](https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-an-order)).
* Verify that there is a new key `_payment_transaction_id` in the response order object's meta_data property with a value starting with 'txn_'.
* Also verify in WP Admin that the transaction id for the above order on the Transactions page matches the value in the previous step.


https://github.com/Automattic/woocommerce-payments/assets/1288616/c5bbc618-0588-40dd-998d-df38314aea8f

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

anu-rock committed:
Streamline the params in attach_intent_info_to_order

The method has too many params whose values can easily be derived from a single intent object.

This commit replaces all derivable params with an intent param.

This will make it easier to add new intent-based data values to order meta.

anu-rock committed:
Save balance transaction id to order meta while processing a payment

anu-rock committed:
Add changelog

anu-rock committed:
Merge branch 'develop' into update/7856-add-charge-txn-id-to-order-meta

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7945` or branch name `update/7856-add-charge-txn-id-to-order-meta` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/7856-add-charge-txn-id-to-order-meta) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: d3d26559c716a7c57fbca802e9fef2e97e26492e
- Build time: 2024-01-18 10:06:55 UTC

_Note: the build is updated when a new commit is pushed to this PR._

anu-rock committed:
Replace the value of balance transaction in create charge helper with a more accurate mock object

anu-rock committed:
Fix psalm error UndefinedMethod

anu-rock committed:
Merge branch 'develop' into update/7856-add-charge-txn-id-to-order-meta

anu-rock commented on includes/class-wc-payments-order-service.php:
```diff
@@ -773,12 +773,12 @@ public function get_fraud_meta_box_type_for_order( $order ) : string {
 	 * Given the payment intent data, adds it to the given order as metadata and parses any notes that need to be added
 	 *
-	 * @param WC_Order                           $order The order.
-	 * @param WC_Payments_API_Abstract_Intention $intent The payment or setup intention object.
```
I was getting Psalm UndefinedMethod linting error with this, so replaced the type hint with a list of possible concrete types (see below). I wonder if there's a proper way to use abstract types as type hints without Psalm complaining.

anu-rock commented on includes/class-wc-payments-order-service.php:
```diff
@@ -773,12 +773,12 @@ public function get_fraud_meta_box_type_for_order( $order ) : string {
-	 * @param WC_Order                           $order The order.
-	 * @param WC_Payments_API_Abstract_Intention $intent The payment or setup intention object.
+	 * @param WC_Order                                                          $order The order.
+	 * @param WC_Payments_API_Payment_Intention|WC_Payments_API_Setup_Intention $intent The payment or setup intention object.
```
I was getting Psalm UndefinedMethod linting error, so replaced the abstract class type hint with a list of possible concrete types. I wonder if there's a proper way to use abstract types as type hints without Psalm complaining.

anu-rock committed:
Update method call in the main gateway class

Looks like new reference of the method that came with the latest code from develop.

dpaun1985 commented on includes/class-wc-payments-order-service.php:
```diff
@@ -752,6 +773,37 @@ public function get_fraud_meta_box_type_for_order( $order ) : string {
+	 *
+	 * @throws Order_Not_Found_Exception
+	 */
+	public function attach_intent_info_to_order( WC_Order $order, $intent ) {
```
part of this method ( calling the setters and calling the save method) is identical to `attach_intent_info_to_order__legacy` .  
This method extracts the data that needs to be added to the order, and saves it on the order.  `attach_intent_info_to_order__legacy` only saves the information to the order. 
We can keep this method to extract the data and call the `attach_intent_info_to_order__legacy` in this method.
So, we could:
1. change this method name to `add_intent_info_to_order`
2. keep the old name for `attach_intent_info_to_order__legacy` ( `attach_intent_info_to_order` )
3. call `attach_intent_info_to_order` in this method

naman03malhotra replied:
I agree with this, common code can be extracted to a separate function as part of this refactor.

anu-rock replied:
@dpaun1985 @naman03malhotra Thanks for the feedback. I fixed the code duplication in https://github.com/Automattic/woocommerce-payments/pull/7945/commits/a7bef878a2b8799e82d98fefb3b0be541a124ed3.

In the long term, I think it's desirable to remove the legacy method and keep only the refactored one. This will allow us to care about only one method when adding new intent properties to order meta.

> change this method name to add_intent_info_to_order

Unfortunately, I cannot change the name as I'd like my changes to apply to all uses of this method.

dpaun1985 commented on includes/class-wc-payments-order-service.php:
```diff
@@ -41,6 +41,13 @@ class WC_Payments_Order_Service {
+	 *
+	 * @const string
+	 */
+	const PAYMENT_TRANSACTION_ID_META_KEY = '_payment_transaction_id';
```
most of the constant values are prefixed with `_wcpay`

anu-rock replied:
Yeah, all the new meta keys seem to follow that pattern. Fixed in https://github.com/Automattic/woocommerce-payments/pull/7945/commits/adc9a67919afb8df6bea59510f30566b49383c5b.

dpaun1985 commented on tests/unit/test-class-wc-payments-order-service.php:
```diff
@@ -1206,13 +1206,9 @@ public function test_get_fraud_meta_box_type() {
-		$currency       = 'USD';
-		$this->order_service->attach_intent_info_to_order( $this->order, $intent_id, $intent_status, $payment_method, $customer_id, $charge_id, $currency );
+		$intent_id = 'pi_mock';
+		$intent    = WC_Helper_Intention::create_intention( [ 'id' => $intent_id ] );
```
since `pi_mock` is used only in once place, it makes no sense to define a variable for it

anu-rock replied:
Good point. Fixed in https://github.com/Automattic/woocommerce-payments/pull/7945/commits/1f4681ac38ffcf9e6eca44eef15927205cc5a9d4.

anu-rock replied:
Whoops, the variable is being used here too:  
https://github.com/Automattic/woocommerce-payments/blob/1f4681ac38ffcf9e6eca44eef15927205cc5a9d4/tests/unit/test-class-wc-payments-order-service.php#L1218

Reverting the change.

dpaun1985 reviewed (commented):
LGTM, however I have some minor remarks. 
I did only code review, I will do the tests after the above are addressed

naman03malhotra commented on includes/class-wc-payments-order-service.php:
```diff
@@ -752,6 +773,37 @@ public function get_fraud_meta_box_type_for_order( $order ) : string {
+		$charge                 = $intent instanceof WC_Payments_API_Payment_Intention ? $intent->get_charge() : null;
+		$charge_id              = $charge ? $charge->get_id() : null;
+		$payment_transaction    = $charge ? $charge->get_balance_transaction() : null;
+		$payment_transaction_id = $payment_transaction ? $payment_transaction['id'] : '';
```
```suggestion
		$payment_transaction_id = $payment_transaction['id'] ?? '';
```
Maybe can be written as ^

mordeth committed:
Merge branch 'develop' into update/7856-add-charge-txn-id-to-order-meta

anu-rock commented:
Some more feedback from Gamma: p1704266984371909-slack-C01BPL3ALGP

anu-rock committed:
Remove code duplication between the two attach intent methods

anu-rock committed:
Prepend wcpay to the meta key

anu-rock committed:
Remove the unnecessary variable in test

anu-rock committed:
Use null coalescing instead of ternary operator

Co-authored-by: Naman Malhotra <naman03malhotra@gmail.com>

anu-rock committed:
Revert "Remove the unnecessary variable in test"

This reverts commit 1f4681ac38ffcf9e6eca44eef15927205cc5a9d4.

dpaun1985 committed:
Merge branch 'develop' into update/7856-add-charge-txn-id-to-order-meta

dpaun1985 reviewed (approved):
✅ LGTM
Tested and the transaction id meta was added on the order.
<img width="296" alt="Screenshot 2024-01-15 at 17 04 20" src="https://github.com/Automattic/woocommerce-payments/assets/82826872/2f637a2c-4854-4234-a8ec-9a7ab98e1589">

anu-rock commented:
@dpaun1985 Thanks for your re-testing and approval. But I wonder why the key is named `_payment_transaction_id` in your screenshot when it should be `_wcpay_payment_transaction_id` as per changes I pushed last week as per your suggestion.

dpaun1985 commented:
> @dpaun1985 Thanks for your re-testing and approval. But I wonder why the key is named `_payment_transaction_id` in your screenshot when it should be `_wcpay_payment_transaction_id` as per changes I pushed last week as per your suggestion.

I think I didn't updated the branch and was an older commit. I just tested and it's the correct key
<img width="325" alt="Screenshot 2024-01-16 at 11 19 19" src="https://github.com/Automattic/woocommerce-payments/assets/82826872/56feba3c-06e3-42b4-bcd6-a1997d254b8a">

anu-rock committed:
Prevent overwriting order meta after successful payment

anu-rock committed:
Add test for the new setter method

anu-rock committed:
Save payment transaction id in the new order service

anu-rock committed:
Fix the undefined variable error

mordeth reviewed (approved):
✅ Code looks good!
✅ Manual checks have been successful.

<img width="317" alt="Screenshot 2024-01-16 at 16 18 57" src="https://github.com/Automattic/woocommerce-payments/assets/8667118/05051a3d-3161-432e-bd94-fbe3e72f0a7f">

anu-rock committed:
Merge branch 'develop' into update/7856-add-charge-txn-id-to-order-meta

anu-rock committed:
Merge branch 'develop' into update/7856-add-charge-txn-id-to-order-meta
